Prometheus backend

shlevy opened the pull request.

shlevy committed:
prometheus: Initial scaffolding

shlevy committed:
prometheus: Initialize selector metrics on backend init

shlevy committed:
prometheus: Allow renderer to emit metric modifications during the event lifecycle

shlevy committed:
prometheus: Break down MetricModification into modifiers for each metric type

shlevy committed:
prometheus: Add metric-specific DSL commands

shlevy committed:
prometheus: Document future work on nested selector types

jhbertra commented on eventuo11y-prometheus/src/Observe/Event/Render/Prometheus.hs:
```diff
@@ -0,0 +1,192 @@
+    pure (gId, g)
+  histograms <- fmap fromAscList . for [minBound @(Histogram es) ..] $ \hId -> do
+    h <- liftIO $ registerHistogram (metricName hId) (metricLabels hId) (metricBounds hId) registry
+    pure (hId, h)
```
hoist `liftIO`

```suggestion
prometheusEventBackend registry render = liftIO $ do
  counters <- fmap fromAscList . for [minBound @(Counter es) ..] $ \cId -> do
    c <- registerCounter (metricName cId) (metricLabels cId) registry
    pure (cId, c)
  gauges <- fmap fromAscList . for [minBound @(Gauge es) ..] $ \gId -> do
    g <- registerGauge (metricName gId) (metricLabels gId) registry
    pure (gId, g)
  histograms <- fmap fromAscList . for [minBound @(Histogram es) ..] $ \hId -> do
    h <- registerHistogram (metricName hId) (metricLabels hId) (metricBounds hId) registry
    pure (hId, h)
```

jhbertra commented on eventuo11y-prometheus/src/Observe/Event/Render/Prometheus.hs:
```diff
@@ -0,0 +1,192 @@
+        emitImmediateEvent = \(NewEventArgs {..}) -> do
+          let PrometheusRendered {..} = render newEventSelector
+          performModifications $ onStart newEventInitialFields Immediate
+          pure PrometheusReference
```
hoist `liftIO`

```suggestion
      performModifications = traverse_ performModification
  pure $
    EventBackend
      { newEvent = \(NewEventArgs {..}) -> liftIO $ do
          let PrometheusRendered {..} = render newEventSelector
          performModifications $ onStart newEventInitialFields Extended
          fieldsRef <- newIORef []
          pure $
            Event
              { reference = PrometheusReference,
                addField = \f -> liftIO $ do
                  performModifications $ onField f
                  atomicModifyIORef' fieldsRef $ \fields ->
                    (f : fields, ()),
                finalize = \e -> liftIO $ do
                  fields <- readIORef fieldsRef
                  performModifications $ onFinalize e (newEventInitialFields ++ (reverse fields))
              },
        emitImmediateEvent = \(NewEventArgs {..}) -> liftIO $ do
          let PrometheusRendered {..} = render newEventSelector
          performModifications $ onStart newEventInitialFields Immediate
          pure PrometheusReference
```

jhbertra reviewed (approved):
LGTM, some minor nitpicks

shlevy committed:
prometheus: Move backend into IO.

Users can liftIO the backend initialization function and
hoistEventBackend the backend themselves.
